gpgsetup.8
.TH GPGSETUP 8 "07 August 2019"
.\"============================
.SH NAME
.\"============================
.
gpgsetup \- open encrypted volumes with GnuPG
.
.\"============================
.SH SYNOPSIS
.\"============================
.SY gpgsetup
.OP \-aAgGvf
.OP \-r name
.OP \-h dir
.OP \-c file
.OP \-m dir
.OP \-\-tmp tmpdir
.OP \-t template
.OP \-u param
.OP \-d dev
.OP \-\-cipher cipher
.OP \-\-offset offset
.OP \-\-skip skip
.OP \-\-postadd command
.OP \-\-prerm command
.RI [ operation ...]
.YS
.
.PP
.BR gpgsetup " [...] [" list ]
.
.br
.BR gpgsetup " [...] " open
.I name
.
.br
.BR gpgsetup " [...] [" \-\-defer "] " close
.I name
.
.br
.BR gpgsetup " [...] [" \-s
.IR size ]
.BI generate " name"
.
.\".br
.\".BR gpgsetup " [...] " insert
.\".I name
.
.br
.BR gpgsetup " [...] " update
.I name
.
.br
.BR gpgsetup " [...] [" \-\-show\-key | \-\-show\-raw "] " show
.I name
.
.br
.BR gpgsetup " [...] [" \-\-show-blob "] " config
.
.br
.BR gpgsetup " [...] " create
.IR name " " dev
.
.\"============================
.SH DESCRIPTION
.\"============================
.B gpgsetup
is used to mount encrypted partitions using key material which have been encrypted using
.BR gpg (1).
.
.PP
See
.BR cryptsetup (8)
for a discussion on encrypted partitions in general, and
.BR gpg (1)
for discussion on keys and using devices such as hardware security modules.
.
.
.SS Operations
The above operations are explained in the following.
.
.TP
.B list
List available blobs to unlock.
.
.TP
.BI "open " name
Unlock the device represented by the key under the name
.IR name .
.
.TP
.BI "close " name
Remove the mapping for the device represented by the name
.IR name .
.
.TP
.BI "generate " name
Generate a blob called
.IR name .
.
.TP
.BI "update " name
Update the blob
.I name
with the given parameters.
.
.TP
.BI "show " name
Show the blob
.IR name .
.
.IP
If
.B \-\-show\-key
is supplied, then the KEY parameter is displayed. Otherwise, the KEY parameter
is hidden.
.
.TP
.B config
Show the currently active configuration parameters.
.
.IP
If
.B \-\-show\-blob
is supplied, then the default blob configuration parameters are shown instead.
.
.TP
.BI "create " name " " dev\c
.RB " (" experimental )
Create a blob from an existing LUKS volume.
.
.I name
is the name of the blob to create and
.I dev
is a path to the LUKS volume.
.
If the LUKS volume has a header on a different device/file, then
.I dev
should be the path of where the header is.
In this case, the actual device path can be set with
.BR \-\-device .
.
.SS Blob format
.B gpgsetup
stores the necessary key material in a dedicated directory.
The material is stored as an environment-style file which is encrypted and
stored in the dedicated directory in a file ending in \(oq.gpg\(cq.
Below is an example of such material before encryption
.PP
.RS
.EX
DEV=/dev/disk/by-uuid/a74c5499-e74a-461f-bbf1-ddf95532309d
KEY=00112233445566778899aabbccddeeff
PRERM=/sbin/vgchange \-an encrypt\-vg
POSTADD=/sbin/vgchange \-ay encrypt\-vg
.EE
.RE
.
.PP
Here's another example:
.PP
.RS
.EX
DEV=/dev/disk/by-partuuid/3aa45469-ae38-4d8b-b0d0-c6203f8a566b
KEY=00112233445566778899aabbccddeeff
CIPHER=aes-xts-plain64
.EE
.RE
.
.PP
The following parameters are supported:
.
.TP
.RB "DEV (" required )
The device node which will be used for dm-crypt.
It is advised to use a symbolic link under \(aq/dev/disk/by-*\(aq
to ensure the named device is consistent.
.
.TP
.RB "KEY (" required )
The key material stored as a lowercase hex string.
The length of this is dependent on the setup, cipher and mode.
.
.TP
POSTADD
A command to run after the device has been unlocked.
.
.TP
PRERM
A command to run before the device is removed.
.
.TP
.RB "CIPHER (" required )
.TQ
OFFSET
.TQ
SKIP
See
.BR cryptsetup (8).
Both OFFSET and SKIP default to zero.
.
.\"============================
.SH OPTIONS
.\"============================
.
.TP
.BR \-r , " \-\-recipient "\c
.I name
encrypt for
.IR name .
defaults to
.B \-\-default\-recipient\-self
in
.BR gpg (1).
.
.TP
.BR \-h , " \-\-homedir "\c
.I dir
.RI "use " dir " as the homedir for
.BR gpg (1).
.
.TP
.BR \-c , " \-\-config "\c
.I file
.RI "use " file " as the configuration file for
.B gpgsetup
instead of the default value of \(oq/etc/gpgsetup.conf\(cq.
.
.TP
.BR \-m , " \-\-material\-dir "\c
.I dir
.RI "use " dir " as the directory containing the key material files."
.
.TP
.BI \-\-tmp " tmpdir"
.RI "use " tmpdir " as the temporary directory.
Defaults to \(oq/tmp\(cq.
.
.TP
.BR \-t , " \-\-template "\c
.I template
.RI "use " template " as a template for the generated blob file."
.
.TP
.BR \-a , " \-\-armour" , " \-\-armor"
use armour on blob (default).
.
.TP
.BR \-A , " \-\-no\-armour" , " \-\-no\-armor"
do not use armour on blob.
.
.\".TP
.\".BR \-k , " \-\-add\-key"
.\"add key to the luks header (default).
.\".
.\".TP
.\".BR \-K , " \-\-no\-add\-key"
.\"do not add key to luks header.
.
.TP
.BR \-s , " \-\-size "\c
.I size
the size of the key to be generated in bytes.
Applies only to the
.B generate
mode of operation.
.
.TP
.BR \-u , " \-\-unset "\c
.I param
unset/unspecify the given parameters.
Valid values include
.BR device ,\: cipher ,\: postadd ,\: prerm ,\: offset ,\: skip .
Used mainly in conjunction with the
.B update
mode of operation.
.
.TP
.BR \-d , " \-\-device "\c
.I dev
the device which is encrypted.
.
.TP
.BI \-\-cipher " cipher"
the cipher specification string.
See
.BR cryptsetup (8)
for details.
.
.TP
.BI \-\-offset " offset"
the offset in 512 byte sectors to start the mapping at.
.
.TP
.BI \-\-skip " skip"
the offset in t512 byte sectors to start in IV calculation.
See
.BR cryptsetup (8)
for details.
.
.TP
.BI \-\-postadd " command"
if present,
.I command
will be executed after setting up the mapping.
.
.TP
.BI \-\-prerm " command"
if present,
.I command
will be executed before removing the mapping.
.
.TP
.BR \-g , " \-\-gpg\-stderr"
show stderr of
.BR gpg (1)
(default).
.
.TP
.BR \-G , " \-\-no\-gpg\-stderr"
do not show stderr of
.BR gpg (1).
.
.TP
.BR \-v , " \-\-verbose"
print additional messages.
.
.TP
.BR \-f , " \-\-force"
force creation of file if already exists.
.
.TP
.BR \-\-defer
defer closing of device.
.
.TP
.B \-\-show\-key
show the key when running
.BR show .
this has no additional effect when using
.BR \-\-show\-raw .
.
.TP
.B \-\-show\-raw
do not parse the blob file, simply print it when running
.BR show .
this shows all, including unknown
options, but also includes the key.
.
.\"============================
.SH BUGS
.\"============================
.
Currently known bugs include:
.
.IP \(bu 3
If
.BR scdaemon (1)
already has ownership of a HSM as a non privileged user, then running
.B gpgsetup
as root will fail.
This can be mitigated by killing the offending
.BR scdaemon (1)
process.
.
.IP \(bu 3
.B gpgsetup
does not currently have a clean way of accessing the
.B \-\-prerm
configuration parameter without creating a temporary,
which ideally wouldn't be necessary.
.RS
.IP \(em 3
This could be solved by having the blob file contain two
sections\(ema plaintext section followed by the encrypted section.
This would require a rewrite of the gpgexec code.
.RE
.
.IP \(bu 3
The
.B libcryptsetup
featured version currently does not support the
.B create
operation, this is likely to change soon.
.
.
.SS Reporting bugs
If you find a bug, please report it to
.MT torin@tcarey.uk
.ME
or preferably, open an issue at
.UR https://www.github.com/torin-carey/gpgsetup
.UE .
.
.\"============================
.SH SEE ALSO
.\"============================
.
.BR gpg (1),
.BR cryptsetup (8)


src/gpgsetup.c
#define _GNU_SOURCE
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <getopt.h>
#include <unistd.h>

#include "gpgsetup.h"
#include "gpgexec.h"
#include "parse.h"

#ifndef CONFIGLOC
#define CONFIGLOC "/etc/gpgsetup.conf"
#endif

const char *usage = // TODO: Update this
"gpgsetup - open LUKS encrypted volumes with GnuPG\n"
"\n"
"gpgsetup [-aAkKgG] [-r name] [-h dir] [-c file] [-m dir] [-t template] [operation...]\n"
"gpgsetup [...] [list]\n"
"gpgsetup [...] open name\n"
"gpgsetup [...] generate name dev [keyslot]\n"
"gpgsetup [...] insert name dev [keyslot]\n"
"gpgsetup [...] [--show-key|--show-raw] show name\n"
"\n"
"-r, --recipient name\n"
"  encrypt for name. defaults to --default-recipient-self in\n"
"  gpg(1).\n"
"\n"
"-h, --homedir dir\n"
"  use dir as the homedir for gpg(1).\n"
"\n"
"-c, --config file\n"
"  use file as the configuration file for gpgsetup instead of the\n"
"  default value of '/etc/gpgsetup.conf'.\n"
"\n"
"-m, --material-dir dir\n"
"  use dir as the directory containing the key material files.\n"
"\n"
"-t, --template template\n"
"  use template as a template for the generated blob file.\n"
"\n"
"-a, --armour, --armor\n"
"  use armour on blob (default).\n"
"\n"
"-A, --no-armour, --no-armor\n"
"  do not use armour on blob.\n"
"\n"
"-k, --add-key\n"
"  add key to the luks header (default).\n"
"\n"
"-K, --no-add-key\n"
"  do not add key to luks header.\n"
"\n"
"-g, --gpg-stderr\n"
"  show stderr of gpg(1) (default).\n"
"\n"
"-G, --no-gpg-stderr\n"
"  do not show stderr of gpg(1).\n"
"\n"
"--show-key\n"
"  show the key when running show. this has no additional effect\n"
"  when using --show-raw.\n"
"\n"
"--show-raw\n"
"  do not parse the blob file, simply print it when running show.\n"
"  this shows all, including unknown options, but also includes the\n"
"  key.\n";



void print_usage(void)
{
	fputs(usage, stderr);
	exit(-1);
}

#define LONG_CIPHER 0
#define LONG_OFFSET 1
#define LONG_SKIP 2
#define LONG_POSTADD 3
#define LONG_PRERM 4
#define LONG_SHOWKEY 5
#define LONG_SHOWRAW 6
#define LONG_TMP 7
#define LONG_SHOWBLOB 8
#define LONG_DEFER 9

const struct option optlong[] = {
	{"cipher", required_argument, NULL, 0},
	{"offset", required_argument, NULL, 0},
	{"skip", required_argument, NULL, 0},
	{"postadd", required_argument, NULL, 0},
	{"prerm", required_argument, NULL, 0},
	{"show-key", no_argument, NULL, 0},
	{"show-raw", no_argument, NULL, 0},
	{"tmp", required_argument, NULL, 0},
	{"show-blob", no_argument, NULL, 0},
	{"defer", no_argument, NULL, 0},

	{"recipient", required_argument, NULL, 'r'},
	{"homedir", required_argument, NULL, 'h'},
	{"armour", no_argument, NULL, 'a'},
	{"armor", no_argument, NULL, 'a'},
	{"no-armour", no_argument, NULL, 'A'},
	{"no-armor", no_argument, NULL, 'A'},
	{"gpg-stderr", no_argument, NULL, 'g'},
	{"no-gpg-stderr", no_argument, NULL, 'G'},
	{"verbose", no_argument, NULL, 'v'},
	{"device", required_argument, NULL, 'd'},
	{"size", required_argument, NULL, 's'},
	{"force", no_argument, NULL, 'f'},
	{NULL, 0, NULL, 0}
};
const char *optstring = "r:h:c:m:t:aAgGvd:s:f";

#define FLAG_SHOWBLOB 1

int main(int argc, char **argv)
{
	struct gpgsetup_config gpgsetup_opt = GPGSETUP_CONF_INITIALISER;
	struct gpgsetup_config gpgsetup_env = GPGSETUP_CONF_INITIALISER;
	struct gpgsetup_config gpgsetup_conf = GPGSETUP_CONF_INITIALISER;

	struct gpgsetup_blob blob_opt = GPGSETUP_BLOB_INITIALISER;
	const char *config_loc = CONFIGLOC;
	int config_spec = 0;
	int flags = 0;

	int r, longind;
	while ((r = getopt_long(argc, argv, optstring, optlong, &longind)) != -1) {
		switch (r) {
		case 0:
			switch (longind) {
			case LONG_CIPHER:
				blob_opt.cipher = optarg;
				blob_opt.specified |= BLOB_CIPHER;
				break;
			case LONG_OFFSET:
				sscanf(optarg, "%zu", &blob_opt.offset);
				blob_opt.specified |= BLOB_OFFSET;
				break;
			case LONG_SKIP:
				sscanf(optarg, "%zu", &blob_opt.skip);
				blob_opt.specified |= BLOB_SKIP;
				break;
			case LONG_POSTADD:
				blob_opt.postadd = optarg;
				blob_opt.specified |= BLOB_POSTADD;
				break;
			case LONG_PRERM:
				blob_opt.prerm = optarg;
				blob_opt.specified |= BLOB_PRERM;
				break;
			case LONG_SHOWKEY:
				gpgsetup_opt.flags |= CONFIG_SHOWKEY;
				gpgsetup_opt.specified |= CONFIG_SHOWKEY;
				break;
			case LONG_SHOWRAW:
				gpgsetup_opt.flags |= CONFIG_SHOWRAW;
				gpgsetup_opt.specified |= CONFIG_SHOWRAW;
				break;
			case LONG_TMP:
				gpgsetup_opt.tmp = optarg;
				gpgsetup_opt.specified |= CONFIG_TMP;
				break;
			case LONG_SHOWBLOB:
				flags |= FLAG_SHOWBLOB;
				break;
			case LONG_DEFER:
				gpgsetup_opt.flags |= CONFIG_DEFER;
				gpgsetup_opt.specified |= CONFIG_DEFER;
				break;
			}
			break;
		case 'r':
			gpgsetup_opt.recipient = optarg;
			gpgsetup_opt.specified |= CONFIG_RECIPIENT;
			break;
		case 'h':
			gpgsetup_opt.homedir = optarg;
			gpgsetup_opt.specified |= CONFIG_HOMEDIR;
			break;
		case 'c':
			config_loc = optarg;
			config_spec = 1;
			break;
		case 'm':
			gpgsetup_opt.materialdir = optarg;
			gpgsetup_opt.specified |= CONFIG_MATERIAL;
			break;
		case 't':
			gpgsetup_opt.templ = optarg;
			gpgsetup_opt.specified |= CONFIG_TEMPLATE;
			break;
		case 'a':
			gpgsetup_opt.flags |= CONFIG_ARMOUR;
			gpgsetup_opt.specified |= CONFIG_ARMOUR;
			break;
		case 'A':
			gpgsetup_opt.flags &= ~CONFIG_ARMOUR;
			gpgsetup_opt.specified |= CONFIG_ARMOUR;
			break;
		case 'g':
			gpgsetup_opt.flags |= CONFIG_GSTDERR;
			gpgsetup_opt.specified |= CONFIG_GSTDERR;
			break;
		case 'G':
			gpgsetup_opt.flags &= ~CONFIG_GSTDERR;
			gpgsetup_opt.specified |= CONFIG_GSTDERR;
			break;
		case 'v':
			gpgsetup_opt.flags |= CONFIG_VERBOSE;
			gpgsetup_opt.specified |= CONFIG_VERBOSE;
			break;
		case 'd':
			blob_opt.dev = optarg;
			blob_opt.specified |= BLOB_DEV;
			break;
		case 's':
			sscanf(optarg, "%zu", &gpgsetup_opt.keysize);
			gpgsetup_opt.specified |= CONFIG_KEYSIZE;
			break;
		case 'f':
			gpgsetup_opt.flags |= CONFIG_FORCE;
			gpgsetup_opt.specified |= CONFIG_FORCE;
			break;
		default:
			print_usage();
			continue;
		}
	}

	enum gpgsetup_mode mode = LIST;
	struct gpgsetup_param param = {NULL, NULL, &blob_opt};

	if (argc > optind) {
		if (!strcmp(argv[optind], "list")) {
//			mode = LIST;
			if (argc - optind != 1)
				print_usage();
		} else if (!strcmp(argv[optind], "open")) {
			mode = OPEN;
			if (argc - optind != 2) {
				fputs("open requires a name\n", stderr);
				print_usage();
			}
		} else if (!strcmp(argv[optind], "close")) {
			mode = CLOSE;
			if (argc - optind != 2) {
				fputs("close requires a name\n", stderr);
				print_usage();
			}
		} else if (!strcmp(argv[optind], "generate")
				|| !strcmp(argv[optind], "gen")) {
			mode = GENERATE;
			if (argc - optind != 2) {
				fputs("generate requires a name\n", stderr);
				print_usage();
			}
		} else if (!strcmp(argv[optind], "update")) {
			mode = UPDATE;
			if (argc - optind != 2) {
				fputs("update requires a name\n", stderr);
				print_usage();
			}
		} else if (!strcmp(argv[optind], "show")) {
			mode = SHOW;
			if (argc - optind != 2) {
				fputs("show requires a name\n", stderr);
				print_usage();
			}
		} else if (!strcmp(argv[optind], "config")) {
			mode = CONFIG;
		} else if (!strcmp(argv[optind], "create")) {
			mode = CREATE;
			if (argc - optind != 3) {
				fputs("create requires a name and device\n", stderr);
				print_usage();
			}
		} else {
			print_usage();
		}
	}
	++optind;

	FILE *conf = fopen(config_loc, "r");
	if (!conf) {
		if (!config_spec)
			goto post_config_load;
		fprintf(stderr, "failed to open configuration file: %m\n");
		return -1;
	}

	if (read_config_file(conf, config_callback, &gpgsetup_conf)) {
		fprintf(stderr, "failed to read configuration file\n");
		return -1;
	}
	fclose(conf);

post_config_load:

	if (read_config_env(config_callback, &gpgsetup_env)) {
		fprintf(stderr, "failed to read environment variables\n");
		return -1;
	}
	apply_config(&gpgsetup_conf, &gpgsetup_env);
	apply_config(&gpgsetup_conf, &gpgsetup_opt);

	switch (mode) {
	case LIST:
		r = handle_mode_list(&gpgsetup_conf, &param);
		break;
	case OPEN:
		param.name = argv[optind++];
		r = handle_mode_open(&gpgsetup_conf, &param);
		break;
	case CLOSE:
		param.name = argv[optind++];
		r = handle_mode_close(&gpgsetup_conf, &param);
		break;
	case GENERATE:
		param.name = argv[optind++];
		r = handle_mode_generate(&gpgsetup_conf, &param);
		break;
	case UPDATE:
		param.name = argv[optind++];
		r = handle_mode_update(&gpgsetup_conf, &param);
		break;
	case SHOW:
		param.name = argv[optind++];
		r = handle_mode_show(&gpgsetup_conf, &param);
		break;
	case CONFIG:
		if (flags & FLAG_SHOWBLOB) {
			struct gpgsetup_blob defaults = GPGSETUP_BLOB_INITIALISER;
			apply_blob_right(&defaults, &blob_opt);
			defaults.specified = -1;
			print_blob(&defaults, stdout, 0);
			//free_blob(&defaults); // Not actually necessary
		} else {
			gpgsetup_conf.specified = -1;
			print_config(&gpgsetup_conf, stdout);
		}
		r = 0;
		break;
	case CREATE:
		param.name = argv[optind++];
		param.dev = argv[optind++];
		if (!(blob_opt.specified & BLOB_DEV)) {
			blob_opt.dev = param.dev;
			blob_opt.specified |= BLOB_DEV;
		}
		r = handle_mode_create(&gpgsetup_conf, &param);
	}

	return r;
}
